In [1]:
import pandas as pd
import numpy as np

ruta='/content/drive/MyDrive/Colab Notebooks/ITESO/Ingeniería de características/P2022_MCD3735A/Módulo 5 - Transformación de variables numéricas/'
df=pd.read_csv(ruta+'sesgo.csv')
df.head(2)

1.0       1.1       1.2       1.4        1.5       2.5       3.0  \
0  4.287419  4.078224  2.410316  1.864664  11.968746 -0.403019  0.473020   
1  6.692587  6.596364 -0.467649  4.196815   9.522961 -0.366849  0.211353   

        0.7       0.6       0.8  ...       0.4      -1.0      -0.8      -0.6  \
0  4.975822  3.037414  2.892491  ...  0.838508  5.541238  4.947706  3.559693   
1  1.000911  1.497043  2.665020  ...  1.213810  4.015426  7.715773  6.496774   

       -0.5       -2.0       -1.5       -1.3       -2.5      -3.0  
0  7.864133  22.867150  22.242351  22.833589  21.672629  2.851895  
1  5.833801  18.040815  22.122988  18.310630  21.197021  3.053828  

[2 rows x 21 columns]

In [ ]:
df['1.0'].skew()

1.0212483134362185

In [ ]:
df['1.5'].skew()

1.5003960156092775

In [ ]:
df['-0.8'].skew()

-0.8098415220779148

In [ ]:
for var in df:
  print(float(var)>0)

In [12]:
# Definir las listas resultantes
funcionesP=[
            lambda x:1/x,
            lambda x:1/np.sqrt(x),
            lambda x:np.log10(x),
            lambda x:np.sqrt(x)
]
funcionesN=[
            lambda x:x**2,
            lambda x:x**3,
            lambda x:x**4,
            lambda x:x**5
]
SesgoN=[[],[],[],[]]
SesgoP=[[],[],[],[]]

In [13]:
for var in df:
  x=df[var]-df[var].min()+1
  xp=[]
  s=[]
  if float(var)>0:
    sxp=SesgoP
    func=funcionesP
  else:
    sxp=SesgoN
    func=funcionesN
  for f in func:
    xp.append(f(x))
    s.append(xp[-1].skew())
  idx=np.argmin(np.abs(s))
  sxp[idx].append((var,s[idx]))

In [14]:
SesgoN

[[('-0.8', -0.04577845391022726),
  ('-0.6', 0.021162133162813665),
  ('-0.5', 0.2707479044230265)],
 [('-1.0', 0.24847620494380152), ('-1.3', -0.1420264342283232)],
 [],
 [('-2.0', -0.15704852764586955),
  ('-1.5', 0.09885238685100839),
  ('-2.5', -0.4427494655413234),
  ('-3.0', -0.7375244021019939)]]

In [15]:
SesgoP

[[('3.0', -0.33248537480368906)],
 [('2.5', -0.17119939610416895)],
 [('1.0', -0.35299940437156885),
  ('1.1', -0.36301124937907797),
  ('1.2', -0.1262171870266795),
  ('1.4', 0.011708969406980576),
  ('1.5', -0.25227818308903044)],
 [('0.7', 0.17743449762614963),
  ('0.6', 0.030021310469718978),
  ('0.8', 0.2071555106270847),
  ('0.5', -0.0803635253412746),
  ('0.4', -0.1255012825645533)]]

En conclusión:

mínimo|máximo|$f$
------|------|----
-2.5|-1.5|$x^5$
>-1.5|<-1.3|$x^4$
-1.3|-1.0|$x^3$
-0.9|-0.5|$x^2$
0.4|0.9|$\sqrt{x}$
1.0|>1.5|$\log_{10}(x)$
$\geq$2.5|<3.0|$\frac{1}{\sqrt{x}}$
|>3.0|$\frac{1}{x}$

Considerando simétrica una distribución con sesgo entre $(-0.44,0.27)$.